Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/polygon_soup_to_polygon_mesh.h

```diff
@@ -91,45 +96,78 @@ class Polygon_soup_to_polygon_mesh
-  *
-  * @tparam Polygon a `std::vector<std::size_t>` containing the indices
-  *         of the points of the polygon face
-  *
-  * @param polygons each element in the vector describes a polygon using the index of the vertices
-  *
-  */
-  template<class Polygon>
-  bool is_polygon_soup_a_polygon_mesh(const std::vector<Polygon>& polygons)
+   * \ingroup PMP_repairing_grp
+   *
+   * returns `true` if the soup of polygons defines a valid polygon
+   * mesh that can be handled by
+   * `CGAL::Polygon_mesh_processing::polygon_soup_to_polygon_mesh()`.
+   * It checks that each edge has at most two incident faces and such an edge
+   * is visited in opposite direction along the two face boundaries,
+   * no polygon has twice the same vertex,
+   * and the polygon soup describes a manifold surface.
+   * This function does not require a range of points as an argument
+   * since the check is purely topological. To each vertex of the mesh
+   * is associated an index that is used in the description of the
+   * boundaries of the polygons provided in `polygons`.
+   *
+   * @tparam PolygonRange a model of the concept `SinglePassRange`
+   * whose value_type is a model of the concept `BidirectionalRange`
+   * whose value_type is `std::size_t`.
+   *
+   * @param polygons each element in the range describes a polygon
+   * using the indices of the vertices.
+   *
+   * @sa `orient_polygon_soup()`
+   */
+  template<class PolygonRange>
+  bool is_polygon_soup_a_polygon_mesh(const PolygonRange& polygons)
   {
-    typedef typename std::iterator_traits<
-              typename Polygon::iterator >::value_type                   V_ID;
-
+    typedef typename boost::range_value<
+      typename boost::range_value<
+        PolygonRange>::type >::type V_ID;
+    typedef typename boost::range_value<
+      PolygonRange>::type           Polygon;
+
+    //check there is no duplicated ordered edge, and
+    //check there is no polygon with twice the same vertex
+    std::vector<V_ID> points;
     std::set< std::pair<V_ID, V_ID> > edge_set;
     BOOST_FOREACH(const Polygon& polygon, polygons)
     {
-      std::size_t nb_edges = polygon.size();
+      std::size_t nb_edges = boost::size(polygon);
       if (nb_edges<3) return false;
-      V_ID prev=polygon.back();
+
+      std::set<V_ID> polygon_vertices;
+      V_ID prev= *--boost::end(polygon);
       BOOST_FOREACH(V_ID id, polygon)
+      {
+        points.push_back(id);
         if (! edge_set.insert(std::pair<V_ID, V_ID>(prev,id)).second )
           return false;
         else
           prev=id;
+
+        if (!polygon_vertices.insert(id).second)
+          return false;//vertex met twice in the same polygon
+      }
     }
 
-    return true;
+    //check manifoldness
+    std::sort(points.begin(), points.end());
+    std::unique(points.begin(), points.end());
+
+    std::vector<Polygon> polygons2(polygons.begin(), polygons.end());
```

Comment:
I have a patch locally to remove the need of filling points. I need to think how we could avoid copying polygons.